Add tolerance in grid even-spacing checking & calculation

zfan001 opened the pull request:
Description: 
  /improver/improver/utilities/spatial.py/calculate_grid_spacing does not have tolerance inside. When I tested on a grid in Australia domain, it reported an error:" coordinate points are not equally spaced".  The grid is actually equally spaced, and the tiny difference is probably due to rounding error.

I added a rtol=1.e-5 in the code, which allowed the function to generate spacing for this kind of case.   
A unit test is added to cover this change.

This change is also required for the new regridding code.

Note by @MoseleyS: The new regridding code is going to use this function to identify the grid spacing on an input cube BEFORE it is regridded.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

benfitzpatrick commented on improver/utilities/spatial.py:
```diff
@@ -99,11 +101,22 @@ def calculate_grid_spacing(
+        diffs_diff = np.diff(diffs)
+        diffs_mean = np.mean(diffs)
+
+        if np.any(abs(diffs_diff) > rtol * diffs_mean):
```
Just looking through to refresh my numpy! If you have negative diffs, would you get a problem from the abs/mean?

zfan001 commented on improver_tests/utilities/test_spatial.py:
```diff
@@ -203,6 +203,22 @@ def test_lat_lon_equal_spacing(self):
         result = calculate_grid_spacing(self.lat_lon_cube, "degrees")
         self.assertAlmostEqual(result, 10.0)
 
+    def test_lat_lon_equal_spacing_with_tolerance(self):
```
added, plus some minor changes. Thanks

MoseleyS commented on improver_tests/utilities/test_spatial.py:
```diff
@@ -199,10 +199,47 @@ def test_axis_keyword(self):
 
+    def test_lat_lon_equal_spacing_with_tolerance(self):
+        """Test grid spacing outputs with lat-lon grid with tolerance"""
+        self.lat_lon_cube.coord("longitude").points = [
+            -19.99999,
+            -10.0,
+            0.0,
+            10.0,
+            20.00001,
+        ]
```
```suggestion
        lon_points = self.lat_lon_cube.coord("longitude").points.copy()
        lon_points[0] += 1e-5
        lon_points[-1] += 1e-5
        self.lat_lon_cube.coord("longitude").points = lon_points
```

zfan001 replied:
I also wanted to do your way at the start.  However, I do not know the value of coordinate in the cube. I have to check it to make sure that 1.e-5 works for rtol. The current way is direct, and the reader immediately sees what we are testing and what the result should be.

MoseleyS replied:
The coordinate points are already implicit here because the expected grid spacing is specified. As you aren't specifying `rtol` in the unit-test, it isn't clear anyway.

MoseleyS replied:
If you really want to keep the hard-coded definitions, can we have them once, in the setUp function perhaps, rather than three times?

MoseleyS commented on improver_tests/utilities/test_spatial.py:
```diff
@@ -199,10 +199,47 @@ def test_axis_keyword(self):
+
+    def test_lat_lon_negative_spacing(self):
+        """Test negative grid spacing outputs with lat-lon grid in degrees"""
+        self.lat_lon_cube.coord("longitude").points = [
+            20.00001,
+            10.0,
+            0.0,
+            -10.0,
+            -19.99999,
+        ]
```
```suggestion
        lon_points = self.lat_lon_cube.coord("longitude").points.copy()
        lon_points[0] += 1e-5
        lon_points[-1] += 1e-5
        self.lat_lon_cube.coord("longitude").points = np.flip(lon_points)
```

benfitzpatrick commented:
Looks good @zfan001! CLI tests pass at the MO.

MoseleyS commented on improver_tests/utilities/test_spatial.py:
```diff
@@ -199,10 +199,47 @@ def test_axis_keyword(self):
+
+    def test_lat_lon_unequal_spacing(self):
+        """Test ValueError for unequal spacing with lat-lon grid"""
+        self.lat_lon_cube.coord("longitude").points = [
+            -19.998,
+            -10.0,
+            0.0,
+            10.0,
+            20.00001,
+        ]
```
```suggestion
        lon_points = self.lat_lon_cube.coord("longitude").points.copy()
        lon_points[0] += 1e-3
        lon_points[-1] += 1e-5
        self.lat_lon_cube.coord("longitude").points = lon_points
```

MoseleyS reviewed (changes requested):
I've made a few suggestions for this PR, but I can't help wondering whether the problem is in the meta-data you are using. Have they been defined accurately enough? It would be far better to get the input meta-data accurate than to "fix" the code here. If you are genuinely using a grid where the spacing cannot be represented within machine precision, then I can see that you would need this change.

zfan001 commented:
Hi Stephen,  this kind of tiny grid spacing difference is common in BOM's ACCESS-** result files. It is basically a upstream issue, and we have no choice but to handle it.   Therefore we need this change.

zfan001 commented on improver/utilities/spatial.py:
```diff
@@ -99,11 +101,22 @@ def calculate_grid_spacing(
+
     if len(diffs) > 1:
-        raise ValueError(
-            "Coordinate {} points are not equally spaced".format(coord.name())
-        )
-    return diffs[0]
+        diffs_diff = np.diff(diffs)
+        diffs_mean = np.mean(diffs)
+
+        if np.any(abs(diffs_diff) > rtol * abs(diffs_mean)):
+            raise ValueError(
+                "Coordinate {} points are not equally spaced".format(coord.name())
+            )
+        # use average spacing for considering tiny difference due to rounding error
+        spacing = (coord.points[-1] - coord.points[0]) / (len(coord.points) - 1.0)
+
+    else:
+        spacing = diffs[0]
```
This suggestion makes  neat on code. On the other hand, the user of the function loses the flexibility of specifying rtol. 
Australia domain grid at BOM has some tiny uneven spacing issue. I prefer to have rtol inside arguments  in case we may be required to use it for tuning up rtol slightly.

MoseleyS replied:
The `np.allclose` method takes `rtol` as an argument, so it can be added when it is found to be needed. I don't want the argument present on this function until something actually uses it. We have regretted putting in functionality into IMPROVER too early in the past, we end up having to maintain something we don't even want.

dmentipl replied:
I think using `np.unique` is a bit strange with floats. I'm wondering if it could be made simpler with:

```python
    diffs = np.diff(coord.points)
    diffs_mean = np.mean(diffs)
    if np.any((diffs - diffs_mean) / diffs_mean > rtol):
        raise ValueError(f"Coordinate {coord.name()} points are not equally spaced")
    else:
        return diffs_mean
```

The `if` checks if any "relative error" between a spacing (each `diffs[i]`) and the mean spacing is greater than the relative tolerance. Otherwise just return the mean, which is the same as `spacing = (coord.points[-1] - coord.points[0]) / (len(coord.points) - 1.0)`.

dmentipl replied:
From @MoseleyS (which was posted as an edit on my comment above):

> I think using np.unique is a bit strange with floats.

It depends what you are trying to do. Up until now, this function has only ever been used on grids that we control and know to be precisely regular. Therefore, this is fine. @zfan001's regridding work is the first time we've tried to use this on external grids. I'd rather not change the behaviour for internal grids as it helps us be confident that our meta-data are as we intend. A quick test shows that `np.allclose(..., rtol=0.)` returns False for machine-precision errors, so having a default value `rtol=0`. rather than my earlier suggestion (`rtol=None`) would be better.

MoseleyS commented:
> 
> 
> Hi Stephen, this kind of tiny grid spacing difference is common in BOM's ACCESS-** result files. It is basically a upstream issue, and we have no choice but to handle it. Therefore we need this change.

I am surprised that this is not corrected by your implementation of StaGE. My main concern is that IMPROVER as it stands is highlighting an inconsistency in your meta-data and these changes here are hiding that rather than fixing the root cause. I think it would be better to regrid your data using the IMPROVER regrid CLI onto a target grid where the inconsistency is not present. If this really is a precision-level issue, the data will not be changed noticeably and then it will be accurately described, saving all sorts of potential problems further down the line.

MoseleyS commented:
Hi @zfan001 
I've spoken to @fionaRust about how this PR interacts with the regridding one and now I understand what you are trying to do!
I would like the resulting code behaviour to remain the same for all the existing uses, and the new tolerant approach to apply only when needed. To achieve this, please can you modify the function so that the default value of `rtol` is `None`, and when `None` precision errors are not tolerated. When a tolerance IS specified this tolerance value is passed in to the `np.allclose` method I proposed.
You will need to specify the `rtol` value when calling this function in your regridding PR, and in the tolerance-checking unit-tests.

I hope this makes sense to you, but I am happy to discuss over Teams tomorrow if that would help. I will be online from 8am UK time (5pm Melbourne time, I think).

dmentipl reviewed (changes requested):
This addition is useful. I've provided a suggestion for how `calculate_grid_spacing` could be simplified further.

zfan001 commented:
Thank @MoseleyS and @dmentipl for good suggestions. I will change the code accordingly.

zfan001 commented:
@MoseleyS  I changed rtol default =0., and call function with option rtol=1.e-5
@dmentipl I removed np.unique. 
Would you please review the code again ?  Thank you for your help.

MoseleyS commented on improver/utilities/spatial.py:
```diff
@@ -100,23 +100,15 @@ def calculate_grid_spacing(
+    diffs = np.diff(coord.points)
+    diffs_mean = np.mean(diffs)
 
+    if np.any(abs(diffs - diffs_mean) > rtol * abs(diffs_mean)):
```
Use the Numpy function, rather than reinventing it:
```suggestion
    if not np.allclose(diffs, diffs_mean), rtol=rtol):
```

zfan001 replied:
Hi Stephen, I looked into https://numpy.org/doc/stable/reference/generated/numpy.allclose.html.  numpy.allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False)'s result is based on "absolute(a - b) <= (atol + rtol * absolute(b)) ". 

With np.any(), I am certain that it is doing what I want ( my personal preference as well).

MoseleyS replied:
From that description, you would get exactly the same result by setting `atol=0.`

zfan001 replied:
Hi Stephen,  I saw that you added atol in the code.  Thanks.

MoseleyS reviewed (changes requested):
Thanks @zfan001. This looks much better now. I'd still prefer to use the Numpy function for the comparison rather than reinventing it.

dmentipl reviewed (dismissed):
Thanks for making changes. I would prefer that you use `np.allclose` as pointed out by @MoseleyS but it doesn't block merging for me.

tjtg commented:
> I am surprised that this is not corrected by your implementation of StaGE. My main concern is that IMPROVER as it stands is highlighting an inconsistency in your meta-data and these changes here are hiding that rather than fixing the root cause. I think it would be better to regrid your data using the IMPROVER regrid CLI onto a target grid where the inconsistency is not present. If this really is a precision-level issue, the data will not be changed noticeably and then it will be accurately described, saving all sorts of potential problems further down the line.

@MoseleyS 
The following items contribute to the issue:
- high resolution NWP model data
- grid coordinates in units of degrees
- longitudes near the date line and/or latitudes near the poles
- coordinates using float32 representation

As an example, the BOM ACCESS-C VicTas domain is 890 x 964 with 0.0135 degree (~1.5km, like UKV) spacing. The domain is located in the south-eastern part of Australia.
```
>>> a = np.linspace(139.0, 151.0015, 890, dtype=np.float32)
>>> np.unique(np.diff(a))
array([0.01348877, 0.01350403], dtype=float32)
```
Using float64 coordinates helps with precision, but doesn't fix the issue:
```
>>> b = np.linspace(139.0, 151.0015, 890, dtype=np.float64)
>>> np.unique(np.diff(b))
array([0.0135, 0.0135])
```
Those two values are very close together and initially print as the same, but they're not exactly the same:
```
>>> np.set_printoptions(precision=16)
>>> np.unique(np.diff(b))
array([0.0134999999999934, 0.0135000000000218])
```
Since there are two unique values present in the array, this will hit the current check for `len(diffs) > 1`.

zfan001 commented:
Hi Stephen,  I have merged your code refactoring into this PR. Your changes are good and neat.  Thanks.

MoseleyS reviewed (approved):
Looks good now. Thanks @zfan001.

Thanks also to @tjtg - I had picked up that information...  eventually...  I really ought to read the information I'm given before I start commenting.
